### ISSUE TITLE
quickly restarting singleplayer on windows sometimes fails to bind port

### ISSUE THREAD

**Issue by wsor4035** [luanti-org/luanti#11832]
Minetest version 5.4.1
Operating system: Windows

Summary:
quickly restarting singleplayer(example relaunching when working on a mod/game) yields error `AsyncErr: Failed to bind socket (port already in use?)` when literally nothing has changed.

![image](https://user-images.githubusercontent.com/24964441/144689787-94ab6719-211a-40e8-9ca9-14284575c839.png)

@kilbith reported similar in minetest-dev irc a while back: https://irc.minetest.net/minetest-dev/2021-11-09#i_5895134

**Comment by wsor4035** [comment]
```
[ConnectionSend]: 12188: Bind failed: No error
ACTION[Main]: World at [FILEPATH]
ACTION[Main]: Server for gameid="farlands_reloaded" listening on 0.0.0.0:51794.
ERROR[Main]: ServerError: AsyncErr: Failed to bind socket (port already in use?)
ACTION[Main]: Server: Shutting down
```

dump at request of sfan5

**Comment by neoh4x0r** [comment]
I remember running into this problem when I was using the debugger in **pcsxr** (a ps1 emulator).

The solution was to ensure that the socket flag **SO_REUSEPORT** was set when calling **setsockopt**.

SO_REUSEPORT: allows a port to be reused
Similarly, SO_REUSEADDR allows an address to be reused.

**For example:**
https://lwn.net/Articles/542629/
```c++
int sfd = socket(domain, socktype, 0);
int optval = 1;
setsockopt(sfd, SOL_SOCKET, SO_REUSEPORT, &optval, sizeof(optval));
bind(sfd, (struct sockaddr *) &addr, addrlen);
```

Probably SO_REUSEPORT should be added to the flags passed to setsockopt in this chunk of code.
It's the only place where I see a socket being created.

https://github.com/minetest/minetest/blob/a106bfd456509b676ccba0ac9bef75c214819028/src/network/socket.cpp#L93-L131

**Comment by sfan5** [comment]
`SO_REUSEPORT` would probably help but Minetest shouldn't *need* to use it, it picks a new random port each time.

**Comment by neoh4x0r** [comment]
> `SO_REUSEPORT` would probably help but Minetest shouldn't _need_ to use it, it picks a new random port each time.
> 
> Edit: What is interesting here is the log message `[ConnectionSend]: 12188: Bind failed: No error`. It obviously doesn't make sense. The code correctly handles errors according to POSIX, but maybe Windows does something different.

The only other thing I can think of is address reuse (eg. SO_REUSEADDR) -- but yes having an error thrown without throwing an "error" is difficult to debug.

**Comment by sfan5** [comment]
I couldn't reproduce this myself but I fixed up some of the error reporting code in `socket.cpp`, please test this build:
[minetest-5.5.0-ws.zip](https://github.com/minetest/minetest/files/7782051/minetest-5.5.0-ws.zip)

**Comment by wsor4035** [comment]
has the same issue still

**Comment by red-001** [comment]
Seems like something that would be expected to happen from time to time when a random port is used? 

```c
	// If using simple singleplayer mode, override
	if (start_data.isSinglePlayer()) {
		start_data.name = "singleplayer";
		start_data.password = "";
		start_data.socket_port = myrand_range(49152, 65535);
	} else {
		g_settings->set("name", start_data.name);
	}
```

I see under normal conditions there are around 30 ports opened in that range on my system. I presume under some types of heavy load that goes up. If there's 30 in that range in use than there should be a ~0.2% chance of a collision on any given connection attempt. If something else was using a bunch of UDP ports that could go up quite a bit. Would hit 1% chance at just 160 open UDP connections. If more websites move to using QUIC I expect this problem to get worse.

Even at 0.2% chance that's not that rare. someone playing the game normally might see it once a year. Someone constantly relaunching the game would see it happen more often.

**Comment by red-001** [comment]
Update on this, it's getting quite bad on Windows, I would imagine because of the move to QUIC

**Comment by sfan5** [comment]
So the whole reason is just that we're stepping on another application's port?
That sounds obvious in hindsight but doesn't explain why nobody hit this on Linux (different behavior for localhost?).

**Comment by sfan5** [comment]
Lazy fix: pick something outside of the usual local port range (< 49152 on windows, < 32768 on my Linux system, [others](https://en.wikipedia.org/wiki/Ephemeral_port)) and call it a day.

Actual fix: Let the server bind to port 0 (meaning the OS assigns one). It will be a bit hairy to communicate this from `ConnectionSendThread` -> `IConnection` -> `Server` -> `Game` to where it's actually needed.

**Comment by red-001** [comment]
I'm working on a fix to pass the socket to the server instead, it seems promising, might slightly impact startup time, but it's a lot less ugly than passing it up the call chain like that

**Comment by red-001** [comment]
> So the whole reason is just that we're stepping on another application's port? That sounds obvious in hindsight but doesn't explain why nobody hit this on Linux (different behavior for localhost?).

I think it might just come down to how many sockets are open, and configurations, and bad luck, I have actually seen this once or twice on desktop link. Windows itself seems to have a bunch of SSDP related sockets open for my system, so can add local network environment to that list as well.

**Comment by neoh4x0r** [comment]
> Actual fix: Let the server bind to port 0 (meaning the OS assigns one). It will be a bit hairy to communicate this from `ConnectionSendThread` -> `IConnection` -> `Server` -> `Game` to where it's actually needed.

Yes, allowing the OS to assign the port is the best option since it will provide one that's not being used.

**Comment by neoh4x0r** [comment]
> I'm working on a fix to pass the socket to the server instead, it seems promising, might slightly impact startup time, but it's a lot less ugly than passing it up the call chain like that

Unless the server is running on the same system as the client (ie. singleplayer), it's not going to know what ports are free.

**Comment by red-001** [comment]
> > I'm working on a fix to pass the socket to the server instead, it seems promising, might slightly impact startup time, but it's a lot less ugly than passing it up the call chain like that
> 
> Unless the server is running on the same system as the client, it's not going to know what ports are free.

I don't really understand this comment, so I will just keep working on that unless you have actionable feedback on this.

**Comment by neoh4x0r** [comment]
> > > I'm working on a fix to pass the socket to the server instead, it seems promising, might slightly impact startup time, but it's a lot less ugly than passing it up the call chain like that
> > 
> > 
> > Unless the server is running on the same system as the client, it's not going to know what ports are free.
> 
> I don't really understand this comment, so I will just keep working on that unless you have actionable feedback on this.

Well, from my understanding of the problem, the issue is assigning a socket to the client/server that is already in-use and that fails.

Sfan's solution is to let the local operating system assign it (which will only pick from free ports), but you are saying that "call chain" is expensive and you want the assignment to be handled at the server-level.

What I've pointed out is that your solution only works if the server is running on the same system, otherwise it won't know what ports are free on the remote client and you would be back to potentially assigning ports which are not free.

In other words, how is your solution obtaining a list of free ports?

**Comment by red-001** [comment]
> > > > I'm working on a fix to pass the socket to the server instead, it seems promising, might slightly impact startup time, but it's a lot less ugly than passing it up the call chain like that
> > > 
> > > 
> > > Unless the server is running on the same system as the client, it's not going to know what ports are free.
> > 
> > 
> > I don't really understand this comment, so I will just keep working on that unless you have actionable feedback on this.
> 
> Well, from my understanding of the problem, the issue is assigning a socket to the client/server that is already in-use and that fails.
> 
> Sfan's solution is to let the local operating system assign it (which will only pick from free ports), but you are saying that "call chain" is expensive and you want the assignment to be handled at the server-level.
> 
> What I've pointed out is that your solution only works if the server is running on the same system, otherwise it won't know what ports are free on the remote client and you would be back to potentially assigning ports which are not free.
> 
> In other words, how is your solution obtaining a list of free ports?

By running on the same system, this is a local loopback issue, hence the mention of singleplayer in the title.

**Comment by neoh4x0r** [comment]
> > In other words, how is your solution obtaining a list of free ports?
> 
> By running on the same system, this is a local loopback issue, hence the mention of singleplayer in the title.

Yes, I mentioned that the client and server must be on the same system for your solution to work.

However,  does the same problem happen when not using singleplayer on Windows? Eg. quickly restarting the client when connected to a remote server? (if this is the case, then the server won't know what ports are free on the client and you would be back to square one).

**Comment by red-001** [comment]
> However, does the same problem happen when not using singleplayer on Windows?

no

**Comment by neoh4x0r** [comment]
> > However, does the same problem happen when not using singleplayer on Windows?
> 
> no

Ok, I see.

**Comment by red-001** [comment]
I would refer you to the PR, you are welcome to leave a review there or test it and see if it works for you, or read the code to understand the change.

**Comment by neoh4x0r** [comment]
> I would refer you to the PR, you are welcome to leave a review there or test it and see if it works for you, or read the code to understand the change.

I looked at the PR and you set socket_port to 0 so the OS assigns a free one -- eg. the client and server, in singleplayer, being restarted before the assigned server port has been freed and, with the PR, another free port is used instead of trying to bind to the previous one.

I'm just left wondering if the same would happen on a remote server if it were restarted before the port had been freed. If this is the case, then it technically extends beyond singleplayer and would need the remote server to use an auto-assigned port that would then need to be announced to clients in some fashion.

In other words, it works for local singlerplayer, but would need more work for remote servers (eg. it's a bug that most people are unlikely to notice under normal conditions since the server might not be restarted enough to catch it).

**Comment by red-001** [comment]
> > I would refer you to the PR, you are welcome to leave a review there or test it and see if it works for you, or read the code to understand the change.
> 
> I looked at the PR and you set socket_port to 0 so the OS assigns a free one -- eg. the client and server, in singleplayer, being restarted before the assigned server port has been freed and, with the PR, another free port is used instead of trying to bind to the previous one.
> 
> I'm just left wondering if the same would happen on a remote server if it were restarted before the port had been freed. If this is the case, then it technically extends beyond singleplayer and would need the remote server to use an auto-assigned port that would then need to be announced to clients in some fashion.
> 
> In other words, it works for local singlerplayer, but would need more work for remote servers (eg. it's a bug that most people are unlikely to notice under normal conditions since the server might not be restarted enough to catch it).

Theoretically but I don't really see how this would help on a dedicated server, how are clients meant to know what port to connect to? how's the server list meant to identify the server? Port forwarding, firewall rules, if something else is listening on that port on the server there isn't much to do.


I don't think this has anything to do with the port not being free'ed yet, it just seems like that, because like I explained if there's a decent number of services in that range you will eventually generate a port that clashes, and you will be more likely to do that if you rapidly restart the game, even if the baseline risk is not driven by the restart count.

**Comment by neoh4x0r** [comment]
> Theoretically but I don't really see how this would help on a dedicated server, how are clients meant to know what port to connect to? how's the server list meant to identify the server? Port forwarding, firewall rules, if something else is listening on that port on the server there isn't much to do.

In theory, this could be handled by having a clearing house/server registry that allows servers to securely, and dynamically, announce their connection details (ip/hostname/and port) and the client launcher simply asks the clearing house for a list, which can be cached by the client for some  period of time (either through a local cache time setting or enforced server-side by the registry).

As far as port forwarding and firewalls go, that's an entirely separate (networking) problem and UPnP can be used by a local device to request that a router open a dynamic port for forwarding.

**Comment by red-001** [comment]
> > Theoretically but I don't really see how this would help on a dedicated server, how are clients meant to know what port to connect to? how's the server list meant to identify the server? Port forwarding, firewall rules, if something else is listening on that port on the server there isn't much to do.
> 
> In theory, this could be handled by having a clearing house/server registry that allows servers to securely, and dynamically, announce their connection details (ip/hostname/and port) and the client launcher simply asks the clearing house for a list, which can be cached by the client for some period of time (either through a local cache time setting or enforced server-side by the registry).
> 
> As far as port forwarding and firewalls go, that's an entirely separate problem and UPnP allows a local device to request that a router open a dynamic port for forwarding.

Well drop us a code review, if that change gets merged it's pretty straight forward to just allow passing port 0 to the server and it should figure everything out and grab itself a dynamic port. The game does not support UPnP AFAIK but you are also free to open a PR for that if you want to. I don't think the server list is likely to be modified to not use the server port as part of it's ID however, in the absence of server GUIDs that's part of how different servers from the same host are disambiguated.

**Comment by neoh4x0r** [comment]
> Well drop us a code review, if that change gets merged it's pretty straight forward to just allow passing port 0 to the server and it should figure everything out and grab itself a dynamic port. The game does not support UPnP AFAIK but you are also free to open a PR for that if you want to. I don't think the server list is likely to be modified to not use the server port as part of it's ID however, in the absence of server GUIDs that's part of how different servers from the same host are disambiguated.

Well, the dynamic server list seems to already be implemented by setting [server_announce to true](https://docs.luanti.org/for-server-hosts/setup/), so the only addition there, if it's not already included, would be to add support to include the server's port.

**Comment by sfan5** [comment]
wtf are we talking about here?

**Comment by neoh4x0r** [comment]
> wtf are we talking about here?

I was wondering about this being a potentially masked bug on remote servers (not singleplayer) where the socket port could fail to bind if the server is restarted fast enough.

The rest was a discussion about how to deal with announcing the randomly assigned port to a client through the server list, if different from the default where someone uses dynamic port mapping through UPnP on the router.

**Comment by red-001** [comment]
I'll be honest I still don't really understand your use case, it's not really related to this issue at all if there's actually issues rebinding the server port during a restart. And it would seem to require quite a lot of changes to support it.

**Comment by red-001** [comment]
> So the whole reason is just that we're stepping on another application's port? That sounds obvious in hindsight but doesn't explain why nobody hit this on Linux (different behavior for localhost?).

I think I figured it out @sfan5 , windows is reserving a bunch of ports for OS use without binding to them, `netsh int ipv4 show excludedportrange protocol=udp` this might be because I am running with HyperV (for virtual secure mode, and Linux guests), which apparently means it grabs even more ports for us by the hypervisor guests.

I was feeling I was missing some ports that were in use. The failure rate seemed much higher than it should be, but figured I was either not measuring it correctly or I was missing when ports were in use.

Not sure if the exact ranges are identifying but it looks like the OS reserved around 6000 ports for it's own use within the range Luanti was trying to pick ports in.

**Comment by neoh4x0r** [comment]
> > So the whole reason is just that we're stepping on another application's port? That sounds obvious in hindsight but doesn't explain why nobody hit this on Linux (different behavior for localhost?).
> 
> I think I figured it out [@sfan5](https://github.com/sfan5) , windows is reserving a bunch of ports for OS use without binding to them, `netsh int ipv4 show excludedportrange protocol=udp` this might be because I am running with HyperV (for virtual secure mode, and Linux guests), which apparently means it grabs even more ports for us by the hypervisor guests.

So, the primary issue is Windows being overly aggressive and reserving  an insane number of ports almost to the point of exhausting the pool. 
 
In other words, you can forget what I said regarding worrying if it was a hidden bug in the server (that people generally don't encounter doing normal use), as it's not an issue with the way Lunati attempts to open/reuse a port.

Thus, the fix/PR for singleplayer should be enough to resolve the issue.

It's also no wonder why we don't see this behavior on Linux.

**Comment by red-001** [comment]
> > > So the whole reason is just that we're stepping on another application's port? That sounds obvious in hindsight but doesn't explain why nobody hit this on Linux (different behavior for localhost?).
> > 
> > 
> > I think I figured it out [@sfan5](https://github.com/sfan5) , windows is reserving a bunch of ports for OS use without binding to them, `netsh int ipv4 show excludedportrange protocol=udp` this might be because I am running with HyperV (for virtual secure mode, and Linux guests), which apparently means it grabs even more ports for us by the hypervisor guests.
> 
> So, the primary issue is Windows being overly aggressive and reserving an insane number of ports almost to the point of exhausting the pool.
> 
> In other words, you can forget what I said regarding worrying if it was a hidden bug in the server (that people generally don't encounter doing normal use), as it's not an issue with the way Lunati attempts to open/reuse a port.
> 
> Thus, the fix/PR for singleplayer should be enough to resolve the issue.
> 
> It's also no wonder why we don't see this behavior on Linux.

it's not really poor OS design since a program should not assume a random port is free, there's APIs to be allocated a port and APIs to check what port you got, but this is getting offtopic again.
